Python/bltinmodule.c

```diff
@@ -2628,36 +2645,86 @@ zip_next(zipobject *lz)
             item = (*Py_TYPE(it)->tp_iternext)(it);
             if (item == NULL) {
                 Py_DECREF(result);
+                if (lz->strict) {
+                    goto check;
+                }
                 return NULL;
             }
             PyTuple_SET_ITEM(result, i, item);
         }
     }
     return result;
+check:
+    if (PyErr_Occurred() && !PyErr_ExceptionMatches(PyExc_StopIteration)) {
+        return NULL;
+    }
```

Comment:
I suggest to always call PyErr_Clear(); here, to clear the current exception, if StopIteration was raised.

Otherwise, tp_iternext is called below with an exception set. tp_iternext can return NULL without setting an exception, but it would "inherit" the exception which was already set.

Remove the PyErr_Clear() below (in `if (i)` block).

Reply:
Yeah, my new test cases were causing asserts to be tripped up because of the unhandled exception, so this *needs* to be done.